### PR TITLE
Fix the advanced inventory volume display (related to #18411, #18288)

### PR DESCRIPTION
**Pull request by pierredavidbelanger**
Also pave the way for user configurable volume units (somehow).
I can push more commits on this PR if you want me to :)

I am still a noob here, this PR needs review ;)

### CODE REVIEW

**Comment by pierredavidbelanger** [comment]
After this PR, the advanced inventory looks like this:
capture d ecran 2016-09-19 a 13 06 43

**Comment by mugling** [comment]
Possibly should be two decimals similar to weight?

**Comment by mugling** [comment]
Try and keep this PR simple. You might take some inspiration from  #18300. In particular don't try to solve the problem of configuration or defaults as that is as yet still undecided.

**Comment by mugling** [comment]
In particular note the functions `round_up` and `to_liter`

**Comment by pierredavidbelanger** [comment]
About this two decimals thing, from what I see in the code, the precision is dynamic:
```
        if( it_weight >= 1000.0 ) {
            if( it_weight >= 10000.0 ) {
                print_color = selected ? hilite( c_red ) : c_red;
                it_weight = 9999.0;
            }
            w_precision = 0;
        } else if( it_weight >= 100.0 ) {
            w_precision = 1;
        } else {
            w_precision = 2;
        }
```

Should we do something like that for volume ? 
And if yes, everywhere we print volumes ?

Also, I think, this column is the only place in the advanced inventory where we print more than 1 decimals precision.

So, I may be too picky with the consistency, but I guess its the weight column that needs to be changed to a fixed one decimal precision :)

**Comment by mugling** [comment]
> So, I may be too picky with the consistency

No, I don't think you're being too picky - just that it's a minefield and I'd be reluctant for one of your initial PR's to get bogged down and you to become disheartened when there is a good chance your otherwise going to at least improve the situation in this PR.

I'm leaning towards suggesting that all interfaces should support a toggle between 1 and 3 decimal places (the latter number chosen as that's what we actually store internally).

**Comment by pierredavidbelanger** [comment]
> Try and keep this PR simple [...] don't try to solve the problem of configuration or defaults [...]

This PR solves nothing, I just created dumb stub functions without real logic in them. You can decide later what you want to do with them.

I just copied exactly what existed for the configurable weight:

```c
const char *volume_units()
{
    return _( "L" );
}
```

and

```c
/**
* Convert volume in ml to units defined by user (currently hardcoded to L)
*/
double convert_volume( int volume )
{
    return volume / 1000.0;
}
```

But, I understand that I mixed the concerns here, I will remove this if you want.

**Comment by mugling** [comment]
Try for either the naive or complete solution but not a mix. I'm not averse to either but be aware the latter is ambitious.

**Comment by pierredavidbelanger** [comment]
Ok, for this PR, I will go for the the naive solution.
So I will revert my changes done in `cata_utility`.
And, I will eventually open a new PR, for the complete solution.

Also, about the "*two decimals similar to weight*", what should I do ?

**Comment by mugling** [comment]
> Also, about the "two decimals similar to weight", what should I do ?

The naive option is two decimals for volume. At least that provides a mitigation for items <100ml volume

File: src/advanced_inv.cpp [hunk 1]
```diff
@@ -254,8 +254,8 @@ void advanced_inventory::print_items( advanced_inventory_pane &pane, bool active
             head = string_format( "%3.1f %s  %.1f %s",
                                   convert_weight( squares[pane.get_area()].weight ),
                                   weight_units(),
-                                  convert_volume( units::to_milliliter( squares[pane.get_area()].volume ) ),
-                                  volume_units() );
+                                  units::to_milliliter( squares[pane.get_area()].volume ) / 1000.0,
```
Comment (mugling) [inline src/advanced_inv.cpp]: `units::to_liter` followed by `round_up`

**Comment by pierredavidbelanger** [comment]
something seems wrong with `round_up( units::to_liter( vol ), 2 )`

I guess `units::to_liter` trims the decimals .. so `round_up` is useless.

Here is an example of this bug with the actual version (without my PR):

cataclysm__dark_days_ahead_-_0_c-18541-g63c7b5f-dirty

So, I will continue my *naive* PR anyways for now, and take for granted that `to_liter` will work correctly eventually.

**Comment by pierredavidbelanger** [comment]
I had to assume that `units::to_liter` does not truncate decimals.
So this PR should **not** be merged before we fix `units::to_liter`.

**Comment by mugling** [comment]
Well spotted. Given liters will almost invariably be fractional perhaps we should perhaps change the template so that it is only possible to return `double`?

**Comment by pierredavidbelanger** [comment]
Indeed, this would render the code more readable, instead of having to use an explicit version of the template.

I am not very good with C++ template though (I usually work with Java or Obj-C).
If you dont mind I will leave you this part :)

You can check this PR #18431 , where I made the same fix everywhere we use `to_litre`.

File: src/advanced_inv.cpp [hunk 2]
```diff
@@ -224,31 +224,38 @@ void advanced_inventory::print_items( advanced_inventory_pane &pane, bool active
...
-        int hrightcol = columns -
-                        to_string( convert_weight( g->u.weight_carried() ) ).length() - 3 - //"xxx.y/"
-                        to_string( convert_weight( g->u.weight_capacity() ) ).length() - 3 - //"xxx.y_"
-                        to_string( g->u.volume_carried() / units::legacy_volume_factor ).length() - 1 - //"xxx/"
-                        to_string( g->u.volume_capacity() / units::legacy_volume_factor ).length() - 1; //"xxx|"
-        nc_color color = c_ltgreen;//red color if overload
-        if( g->u.weight_carried() > g->u.weight_capacity() ) {
+        const double weight_carried = convert_weight( g->u.weight_carried() );
+        const double weight_capacity = convert_weight( g->u.weight_capacity() );
+        const char *weight_unit = weight_units();
```
Comment (mugling) [inline src/advanced_inv.cpp]: Unnecessary variable. You use a mix of `weight_unit` and `weight_units()`. Prefer the function

File: src/advanced_inv.cpp [hunk 3]
```diff
@@ -224,31 +224,38 @@ void advanced_inventory::print_items( advanced_inventory_pane &pane, bool active
...
+        const double volume_carried = round_up( units::to_liter( g->u.volume_carried() ), 1 );
+        const double volume_capacity = round_up( units::to_liter( g->u.volume_capacity() ), 1 );
+        const char *volume_unit = _( "L" );
```
Comment (mugling) [inline src/advanced_inv.cpp]: Add this to the format strings. I agree we should later offer choice of units/precision but for now it implies conditional output where there is none.

File: src/advanced_inv.cpp [hunk 4]
```diff
@@ -224,31 +224,38 @@ void advanced_inventory::print_items( advanced_inventory_pane &pane, bool active
...
+        // align right, so calculate formated head length 
+        const std::string head = string_format( "%.1f/%.1f %s  %.1f/%.1f %s",
+                                                weight_carried, weight_capacity, weight_unit,
+                                                volume_carried, volume_capacity, volume_unit );
+        const int hrightcol = columns - 1 - head.length();
+        nc_color color = c_ltgreen; //red color if overload
```
Comment (mugling) [inline src/advanced_inv.cpp]: Ideally candidate for a ternary operator
```
nc_color color =  weight_carried > weight_capacity ? c_red : c_ltgreen;
```
Also would avoid need for comment

File: src/advanced_inv.cpp [hunk 5]
```diff
@@ -259,7 +266,12 @@ void advanced_inventory::print_items( advanced_inventory_pane &pane, bool active
             } else {
                 maxvolume = g->m.max_volume( s.pos );
             }
-            head = string_format( "%3.1f %3d/%3d", convert_weight( s.weight ), s.volume / units::legacy_volume_factor, maxvolume / units::legacy_volume_factor );
+            head = string_format( "%3.1f %s  %.1f/%-.1f %s",
```
Comment (mugling) [inline src/advanced_inv.cpp]: Previously AIM displayed to 2dp. It may be best to preserve this for now (both weight and volume)

Comment (pierredavidbelanger) [inline src/advanced_inv.cpp]: I am sorry, I do not understand this one.
Are you talking about the precision of the weight ?

**Comment by pierredavidbelanger** [comment]
Thank you for your comments by the way, it really helps me better understand the code.

File: src/advanced_inv.cpp [hunk 6]
```diff
@@ -361,13 +362,22 @@ void advanced_inventory::print_items( advanced_inventory_pane &pane, bool active
...
-        if( it_vol > 9999 ) {
-            it_vol = 9999;
-            print_color = selected ? hilite( c_red ) : c_red;
+        if( it_vol >= 100 ) {
+            if( it_vol >= 10000 ) {
+                it_vol = 9999;
+                print_color = selected ? hilite( c_red ) : c_red;
+            }
+            v_precision = 0;
+        } else if( it_vol >= 10 ) {
```
Comment (mugling) [inline src/advanced_inv.cpp]: This causes problems with alignment of the decimal point when you have for example `13.94` and `100.6`. We should eventually support toggling of displayed precision, possibly with automatic default, but I'd avoid such logic here for now

**Comment by mugling** [comment]
Everything else apart from the one new review item checks out

File: src/advanced_inv.cpp (hunk 6 above)
Comment (pierredavidbelanger) [inline src/advanced_inv.cpp]: My thinking behind this change is: the `vol` column only has 4 digits worth of space .. if we hardcode `%4.2f` , all volumes `>= 99.99` will make this column overflow to the left. 

So I had to make this code work [the same as for the weight](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/advanced_inv.cpp#L350): dynamic precision, and thus the weight shows the exact same problem you mention, see this capture from the latest master (without this PR):
cataclysm__dark_days_ahead_-_0_c-18571-g723cee4

Unless there is an option I overlooked, you have to choose between:
- hardcoded precision (2 decimals), for better readability, but may overflow to the left and display useless `0` decimals
- dynamic precision for a better display precision and for consistency with the weight col I guess, but with the decimal point alignment problem you mention

Its up to you.

File: src/advanced_inv.cpp (hunk 5 above)
Comment (pierredavidbelanger) [inline src/advanced_inv.cpp]: So, I assume this one is ok ?

File: src/advanced_inv.cpp (hunk 6 above)
Comment (mugling) [inline src/advanced_inv.cpp]: > My thinking behind this change is: the vol column only has 4 digits worth of space
> will make this column overflow to the left. 

This is a good point and something I hadn't considered. What you have is for now an improvement (and thanks for identifying #18430 in the process) so lets go with this for now

**Pull request merged**